plugins/calibre.koplugin/search.lua

```diff
@@ -0,0 +1,494 @@
+local CalibreMetadata = require("metadata")
+local CenterContainer = require("ui/widget/container/centercontainer")
+local ConfirmBox = require("ui/widget/confirmbox")
+local DataStorage = require("datastorage")
+local Device = require("device")
+local DocumentRegistry = require("document/documentregistry")
+local Font = require("ui/font")
+local InputDialog = require("ui/widget/inputdialog")
+local InfoMessage = require("ui/widget/infomessage")
+local InputContainer = require("ui/widget/container/inputcontainer")
+local LuaSettings = require("luasettings")
+local Menu = require("ui/widget/menu")
+local Screen = require("device").screen
+local UIManager = require("ui/uimanager")
+local logger = require("logger")
+local util = require("util")
+local _ = require("gettext")
+local T = require("ffi/util").template
+
+-- get root dir for disk scans
+local function getDefaultRootDir()
+    if Device:isAndroid() then
+        return Device.external_storage()
+    elseif Device:isDesktop() then
+        return os.getenv("HOME")
+    elseif Device:isKindle() then
+        return "/mnt/us/documents"
+    elseif Device:isRemarkable() then
+        return "/home/root"
+    elseif Device:isKobo() or Device:isCervantes() then
+        return "/mnt"
+    else
+        return "."
+    end
+end
+
+-- get books that exactly match the search tag
+local function getBooksByTag(t, tag)
+    local result = {}
+    for _, book in ipairs(t) do
+        for __, _tag in ipairs(book.tags) do
+            if tag == _tag then
+                table.insert(result, book)
+            end
+        end
+    end
+    return result
+end
+
+-- get books that exactly match the search series
+local function getBooksBySeries(t, series)
+    local result = {}
+    for _, book in ipairs(t) do
+        if type(book.series) ~= "function" then
+            if book.series == series then
+                table.insert(result, book)
+            end
+        end
+    end
+    return result
+end
+
+-- get tags that partially match the search criteria and their frequency
+local function searchByTag(t, query, case_sensitive)
+    local freq = {}
+    for _, book in ipairs(t) do
+        for __, tag in ipairs(book.tags) do
+            -- case insensitive
+            if not query or (query and string.find(
+                string.upper(tag), string.upper(query))) then
+                freq[tag] = (freq[tag] or 0) + 1
+            end
+        end
+    end
+    return freq
+end
+
+-- get series that partially match the search criteria and their frequency
+local function searchBySeries(t, query, case_sensitive)
+    local freq = {}
+    for _, book in ipairs(t) do
+        if type(book.series) ~= "function" then
+            if not query or (query and string.find(
+                string.upper(book.series), string.upper(query))) then
+                freq[book.series] = (freq[book.series] or 0) + 1
```

Comment:
Also not fond when the `if` continuation is at the same level as the then action.
```lua
if something and
    other thing then
    dosomestuff
end
```
I'd rather see more indentation on the if continuation:
```lua
if something and
        other thing then
    dosomestuff
end
```
or
```lua
if something
        and other thing then
    dosomestuff
end
```
Makes it easier to read.

Reply:
Will take your suggestion for next commits. Thanks!